Node admission/authorization

liggitt opened the pull request:
Enables the node authorizer and NodeRestriction admission plugins

liggitt commented:
[test]

liggitt commented:
Will wait for the 1.7 rebase and drop all the upstream commits

liggitt commented:
cc @openshift/sig-security

simo5 commented on pkg/cmd/server/bootstrappolicy/policy.go:
```diff
@@ -966,9 +966,15 @@ func GetOpenshiftBootstrapClusterRoleBindings() []rbac.ClusterRoleBinding {
+				Kind:     "ClusterRole",
+				Name:     NodeRoleName,
+			},
+		},
```
please use newOriginClusterBinding() or even better rbac.newClusterBinding() if you do not have to maintain a name difference between the role and the binding.

liggitt replied:
that helper does not allow empty subjects, which we want here.

simo5 replied:
very annoying, what is the purpose of a binding with no subjects ?

liggitt replied:
so that tightening reconciliation removes the grant to the `system:nodes` group

simo5 commented on pkg/cmd/server/bootstrappolicy/policy_test.go:
```diff
@@ -176,4 +180,19 @@ func TestCovers(t *testing.T) {
+		for _, r := range miss {
+			t.Logf("\t%s", r.CompactString())
+		}
+	}
```
Why is this needed ?

deads2k commented on pkg/cmd/server/origin/master_config.go:
```diff
@@ -779,14 +777,25 @@ func buildKubeAuth(r rbacinformers.Interface) (kauthorizer.Authorizer, rbacregis
 }
 
-func newAuthorizer(kubeAuthorizer kauthorizer.Authorizer, kubeSubjectLocator rbacauthorizer.SubjectLocator, clusterRoleGetter rbaclisters.ClusterRoleLister, projectRequestDenyMessage string) (kauthorizer.Authorizer, authorizer.SubjectLocator) {
+func newAuthorizer(kubeAuthorizer kauthorizer.Authorizer, kubeSubjectLocator rbacauthorizer.SubjectLocator, informers InformerAccess, projectRequestDenyMessage string) (kauthorizer.Authorizer, authorizer.SubjectLocator) {
```
We're trying to kill the big one.  Plumb through the ones you want or make this a method on the config type or make a `Complete` method and we can try to start converging.

liggitt replied:
so plumbed

liggitt commented:
/retest

liggitt commented:
Router flake
/retest

liggitt commented:
any other comments?

enj commented on pkg/cmd/server/bootstrappolicy/policy.go:
```diff
@@ -966,9 +966,15 @@ func GetOpenshiftBootstrapClusterRoleBindings() []rbac.ClusterRoleBinding {
-		newOriginClusterBinding(NodeRoleBindingName, NodeRoleName).
-			Groups(NodesGroup).
-			BindingOrDie(),
+		// Preserve the empty node binding with no subjects for tightening reconciliation
```
Seems like we should have a `GetDeadClusterRoleBindings`.

liggitt replied:
I wouldn't want to chance someone calling this and missing calling that

enj replied:
We already trust callers to use `GetBootstrapClusterRoles` which gives them `GetDeadClusterRoles`.  I would expect `GetBootstrapClusterRoleBindings` to have the same logic.

liggitt replied:
ah, ok

liggitt replied:
created GetDeadClusterRoleBindings

liggitt commented:
updated, and added an integration test that replicates the checks in https://github.com/kubernetes/kubernetes/blob/release-1.7/test/integration/auth/node_test.go against our wired API server to ensure our default authz/admission set up protects us.

enj commented on vendor/k8s.io/kubernetes/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go:
```diff
@@ -405,10 +405,12 @@ func ClusterRoleBindings() []rbac.ClusterRoleBinding {
-		// If an admin wants to grant the system:node role (which cannot partition Node API access), they will need to create their own clusterrolebinding.
-		// TODO: Remove the subjects from this binding in 1.8 (leave the empty binding for tightening reconciliation), and remove AddClusterRoleBindingFilter()
-		rbac.NewClusterBinding(systemNodeRoleName).Groups(user.NodesGroup).BindingOrDie(),
+		// This default binding of the system:node role to the system:nodes group is deprecated in 1.7 with the availability of the Node authorizer.
```
Uhh upstream commit?

liggitt replied:
Yes, pushed too fast. Will fix

liggitt commented:
comments addressed, PTAL

enj commented:
/lgtm
